File: apps/ios/Loonext/Core/Model/Billing.swift
import Foundation

/// POST /v1/billing/checkout and /portal — open in an EXTERNAL browser.
struct HostedUrl: Codable, Sendable {
    let url: String
}

struct UsageMonth: Codable, Sendable {
    let month: String
    let segments: Int
}

struct UsageStorage: Codable, Sendable {
    @Default<DefaultZero> var attachments_bytes: Int
    @Default<DefaultZero> var mms_bytes: Int
    /// Media a customer sent us.
    @Default<DefaultZero> var received_media_bytes: Int
    /// Media we sent out.
    @Default<DefaultZero> var sent_media_bytes: Int
    /// Voicemail recordings we keep in our own bucket.
    @Default<DefaultZero> var voicemail_bytes: Int
    /// Anything stored that the named kinds do not account for.
    @Default<DefaultZero> var other_bytes: Int
    /// Every byte this workspace holds, measured from the buckets themselves.
    @Default<DefaultZero> var total_bytes: Int

    /// What is really stored. The old line added two figures together and so
    /// left voicemail recordings out entirely; the server measures the buckets
    /// now, and the sum is only a fallback for a response that predates it.
    var totalStored: Int {
        total_bytes > 0 ? total_bytes : attachments_bytes + mms_bytes
    }

    // The storage budgets went away with #121 (storage is free and capless), so
    // the fields went with them rather than lingering as zeros nothing reads.
    init(
        attachments_bytes: Int = 0,
        mms_bytes: Int = 0,
        received_media_bytes: Int = 0,
        sent_media_bytes: Int = 0,
        voicemail_bytes: Int = 0,
        other_bytes: Int = 0,
        total_bytes: Int = 0
    ) {
        self.attachments_bytes = attachments_bytes
        self.mms_bytes = mms_bytes
        self.received_media_bytes = received_media_bytes
        self.sent_media_bytes = sent_media_bytes
        self.voicemail_bytes = voicemail_bytes
        self.other_bytes = other_bytes
        self.total_bytes = total_bytes
    }
}

struct UsageVoice: Codable, Sendable {
    @Default<DefaultZero> var used_minutes: Int
    @Default<DefaultZero> var included_minutes: Int
    let cap_minutes: Int?
    @Default<DefaultZero> var overage_minutes: Int
    @Default<DefaultZero> var projected_overage_cents: Int
    @Default<DefaultTrue> var overage_billed: Bool

    init(
        used_minutes: Int = 0,
        included_minutes: Int = 0,
        cap_minutes: Int? = nil,
        overage_minutes: Int = 0,
        projected_overage_cents: Int = 0,
        overage_billed: Bool = true
    ) {
        self.used_minutes = used_minutes
        self.included_minutes = included_minutes
        self.cap_minutes = cap_minutes
        self.overage_minutes = overage_minutes
        self.projected_overage_cents = projected_overage_cents
        self.overage_billed = overage_billed
    }
}

struct UsageOverageProjection: Codable, Sendable {
    @Default<DefaultFalse> var trending_over: Bool
    @Default<DefaultZero> var projected_overage_cents: Int

    init(trending_over: Bool = false, projected_overage_cents: Int = 0) {
        self.trending_over = trending_over
        self.projected_overage_cents = projected_overage_cents
    }
}

enum DefaultEmptyStorage: DefaultCodableProvider {
    static var defaultValue: UsageStorage { UsageStorage() }
}

enum DefaultEmptyVoice: DefaultCodableProvider {
    static var defaultValue: UsageVoice { UsageVoice() }
}

enum DefaultEmptyProjection: DefaultCodableProvider {
    static var defaultValue: UsageOverageProjection { UsageOverageProjection() }
}

/// #178: the fair-use presentation contract. GET /v1/usage derives `status`
/// server-side so every client renders the same philosophy: 'quiet' shows no
/// meters anywhere, 'pacing' shows the early warning, 'capped' shows the
/// owner-set spending cap approaching or reached. Server string-enum, so a
/// lagging build never crashes; unknown values render the calm 'quiet' state.
enum UsageStatus {
    static let quiet = "quiet"
    static let pacing = "pacing"
    static let capped = "capped"
}

/// #178 decode default: keeps pre-#178 cached payloads (and unknown values)
/// decoding as the calm state.
enum DefaultUsageStatusQuiet: DefaultCodableProvider {
    static var defaultValue: String { UsageStatus.quiet }
}

/// GET /v1/usage — nils when the company has never checked out.
struct Usage: Codable, Sendable {
    /// #178 presentation status; the default keeps pre-#178 payloads decoding
    /// as the calm state (unknown values also render quiet).
    @Default<DefaultUsageStatusQuiet> var status: String
    let period_start: String?
    let period_end: String?
    @Default<DefaultZero> var included_segments: Int
    @Default<DefaultZero> var used_segments: Int
    @Default<DefaultZero> var inbound_segments: Int
    @Default<DefaultZero> var overage_segments: Int
    let cap_segments: Int?
    @Default<DefaultZero> var projected_overage_cents: Int
    @Default<DefaultEmptyProjection> var overage_projection: UsageOverageProjection
    @Default<DefaultEmptyList<UsageMonth>> var history: [UsageMonth]
    @Default<DefaultEmptyStorage> var storage: UsageStorage
    @Default<DefaultEmptyVoice> var voice: UsageVoice
    /// What Lou has done this month, per feature. Empty against a server that
    /// predates it, so the section simply does not render.
    @Default<DefaultEmptyList<AiFeatureUsage>> var ai: [AiFeatureUsage]
    /// #426: carrier-reported delivery for the period. Nil against a server
    /// that predates it, or when the read failed — the page still renders.
    var delivery: UsageDelivery? = nil

    init(
        status: String = UsageStatus.quiet,
        period_start: String? = nil,
        period_end: String? = nil,
        included_segments: Int = 0,
        used_segments: Int = 0,
        inbound_segments: Int = 0,
        overage_segments: Int = 0,
        cap_segments: Int? = nil,
        projected_overage_cents: Int = 0,
        overage_projection: UsageOverageProjection = UsageOverageProjection(),
        history: [UsageMonth] = [],
        storage: UsageStorage = UsageStorage(),
        voice: UsageVoice = UsageVoice(),
        ai: [AiFeatureUsage] = []
    ) {
        self.status = status
        self.period_start = period_start
        self.period_end = period_end
        self.included_segments = included_segments
        self.used_segments = used_segments
        self.inbound_segments = inbound_segments
        self.overage_segments = overage_segments
        self.cap_segments = cap_segments
        self.projected_overage_cents = projected_overage_cents
        self.overage_projection = overage_projection
        self.history = history
        self.storage = storage
        self.voice = voice
        self.ai = ai
    }
}

/// One AI feature's month: what has been used against its limit.
struct AiFeatureUsage: Codable, Sendable, Identifiable {
    /// The ledger key, so a row is identified without matching on copy.
    let key: String
    let label: String
    @Default<DefaultZero> var used: Int
    @Default<DefaultZero> var cap: Int
    @Default<DefaultTrue> var enabled: Bool
    /// #431 ask 3 — what people did with the output, beside what it cost.
    ///
    /// Labelled by the server in each feature's own words ("sent as written",
    /// "cleared") so all three clients say the same thing. EMPTY until outcomes
    /// arrive, and an empty list must render as "not measured yet" rather than as
    /// zeroes: a feature used forty times with nothing recorded is an
    /// instrumentation gap, and "0 sent as written" would report that gap as a
    /// verdict on the quality.
    @Default<DefaultEmptyList<AiOutcomeLine>> var outcomes: [AiOutcomeLine]
    /// How many outcomes those lines cover. Separate from `used` because they
    /// will not match — a draft offered and never read is a request with no
    /// outcome — and no rate is computed anywhere, deliberately.
    @Default<DefaultZero> var outcomesRecorded: Int

    var id: String { key }
}

/// #431: what a person did with one feature's output, ready to render.
struct AiOutcomeLine: Codable, Sendable, Identifiable {
    let label: String
    @Default<DefaultZero> var count: Int

    var id: String { label }
}

/// GET /v1/billing/modules — admin-only add-on catalog with enabled state.
struct BillingModules: Codable, Sendable {
    @Default<DefaultEmptyList<BillingModule>> var modules: [BillingModule]
}

struct BillingModule: Codable, Sendable {
    let id: String
    let label: String
    let blurb: String
    let detail: String?
    let monthly_cents: Int
    @Default<DefaultFalse> var enabled: Bool
    /// #41: deliverable AND priced in this environment; refuse to sell otherwise.
    @Default<DefaultFalse> var available: Bool
}

/// POST /v1/billing/change-plan result.
///
/// `reinstated` / `held` arrive with #523: an upgrade raises the allowance, and
/// the API claims against the new one in the same call — so a switch to Pro can
/// bring numbers back, and the client is told which rather than having to
/// refetch and diff.
struct ChangePlanResult: Codable, Sendable {
    let plan: String
    let effective: String
    let effective_at: String?
    @Default<DefaultEmptyList<HeldNumber>> var reinstated: [HeldNumber]
    @Default<DefaultEmptyList<HeldNumber>> var held: [HeldNumber]
}

// MARK: - Numbers the plan does not cover (#523)

/// One number this workspace holds that its plan does not currently cover.
///
/// It is NOT released and it is not gone: the row is intact, texts and calls
/// still land on it, and its history is untouched. What it cannot do is send or
/// answer. The whole point of the surface this feeds is that the state is
/// visible and has a way out, because before #523 a resubscribe onto a smaller
/// plan un-suspended everything and we paid the carrier rent forever.
struct HeldNumber: Codable, Sendable, Identifiable {
    let id: String
    let number_e164: String?
    /// When the hold started. Decoded because it is part of the contract, and
    /// deliberately not rendered — see `HeldNumbersCard`.
    var suspended_at: String? = nil

    init(id: String, number_e164: String?, suspended_at: String? = nil) {
        self.id = id
        self.number_e164 = number_e164
        self.suspended_at = suspended_at
    }
}

/// Why a workspace's numbers are suspended — the server's word for it, not the
/// client's inference.
///
/// The two states look identical in `phone_numbers.status` and mean opposite
/// things: one is "your plan is smaller than your workspace", the other is "your
/// subscription is over and the 30-day hold is running". `GET
/// /v1/billing/held-numbers` decides between them so three clients do not each
/// derive it from two fields and describe the same state three ways.
enum HeldNumbersReason {
    static let overPlanAllowance = "over_plan_allowance"
    static let subscriptionInactive = "subscription_inactive"
}

/// GET /v1/billing/held-numbers (#523) — what this workspace holds beyond what
/// its plan covers, and both ways back.
///
/// EVERY FIGURE THIS SCREEN PRINTS IS IN HERE. The allowance, the plan's hard
/// cap, the price of buying one back and the currency that price is denominated
/// in are all served rather than derived, because a client that renders "$5" out
/// of its own head at a workspace billed in CAD is #522 happening again.
struct HeldNumbers: Codable, Sendable {
    /// "starter" | "pro", or nil when the workspace has never checked out.
    let plan: String?
    /// Numbers the plan itself covers. Nil when `plan` is.
    let included: Int?
    /// Extra numbers actually billed on the subscription right now.
    @Default<DefaultZero> var paid_extras: Int
    /// `included + paid_extras` — what may be active at once.
    let allowance: Int?
    /// The plan's hard TOTAL cap (#80), or nil when it has none (Pro).
    let max_total: Int?
    /// `HeldNumbersReason`, or nil when nothing is held.
    let reason: String?
    @Default<DefaultEmptyList<HeldNumber>> var held: [HeldNumber]
    /// What buying capacity for ONE held number costs, from the price book.
    let extra_number_cents: Int?
    /// The currency `extra_number_cents` is denominated in. OPTIONAL and with
    /// no default: an absent currency means the figure cannot be labelled
    /// honestly, and an unlabelled price is the defect #522 was.
    let extra_number_currency: String?
    /// Whether POST …/reinstate would be accepted right now. Served so the
    /// button can be ABSENT rather than fail — being told "no" after pressing
    /// it is how somebody concludes the product is broken.
    @Default<DefaultFalse> var can_reinstate: Bool
    /// Starter only: the other way back, and it buys no extra number.
    @Default<DefaultFalse> var can_upgrade: Bool

    init(
        plan: String? = nil,
        included: Int? = nil,
        paid_extras: Int = 0,
        allowance: Int? = nil,
        max_total: Int? = nil,
        reason: String? = nil,
        held: [HeldNumber] = [],
        extra_number_cents: Int? = nil,
        extra_number_currency: String? = nil,
        can_reinstate: Bool = false,
        can_upgrade: Bool = false
    ) {
        self.plan = plan
        self.included = included
        self.paid_extras = paid_extras
        self.allowance = allowance
        self.max_total = max_total
        self.reason = reason
        self.held = held
        self.extra_number_cents = extra_number_cents
        self.extra_number_currency = extra_number_currency
        self.can_reinstate = can_reinstate
        self.can_upgrade = can_upgrade
    }
}

/// POST /v1/billing/held-numbers/:id/reinstate (#523).
///
/// THREE OUTCOMES, and they are not the same sentence:
///
///   `reinstated`                  paid, and the number is back.
///   `already_active`              nothing was bought — it was already back.
///                                 A double-press, or an upgrade beat us to it.
///   neither                       the charge landed and the un-hold did not
///                                 (the #110 raise fence refused a capacity
///                                 raise formed against a stale epoch). The
///                                 caller must not invite an immediate retry:
///                                 the money has moved.
struct ReinstatedNumber: Codable, Sendable {
    @Default<DefaultFalse> var reinstated: Bool
    @Default<DefaultFalse> var already_active: Bool
    let paid_extras: Int?
    let allowance: Int?
    @Default<DefaultEmptyList<HeldNumber>> var held: [HeldNumber]

    init(
        reinstated: Bool = false,
        already_active: Bool = false,
        paid_extras: Int? = nil,
        allowance: Int? = nil,
        held: [HeldNumber] = []
    ) {
        self.reinstated = reinstated
        self.already_active = already_active
        self.paid_extras = paid_extras
        self.allowance = allowance
        self.held = held
    }
}

/// #426 — carrier-reported delivery, split by where the message was going.
///
/// The NAME is load-bearing: a receipt means a carrier acknowledged handoff,
/// not that a person read it, so every surface says "carrier-reported".
struct UsageDeliveryCountry: Codable, Sendable {
    /// "US" | "CA" | "other", from the destination's area code.
    @Default<DefaultOtherCountry> var country: String
    @Default<DefaultZero> var delivered: Int
    @Default<DefaultZero> var failed: Int
    /// Accepted by us, not yet acknowledged by a carrier. Not a failure.
    @Default<DefaultZero> var pending: Int
    /// delivered / (delivered + failed), or NIL when too few have settled to
    /// mean anything. Render counts and never a percentage when nil: one
    /// failure out of forty reads as 2.5% and usually means a disconnected
    /// number, which is manufactured worry rather than information.
    let rate: Double?
}

struct UsageDelivery: Codable, Sendable {
    @Default<DefaultEmptyList<UsageDeliveryCountry>> var by_country: [UsageDeliveryCountry]
    @Default<DefaultZero> var delivered: Int
    @Default<DefaultZero> var failed: Int
    @Default<DefaultZero> var pending: Int
}

enum DefaultOtherCountry: DefaultCodableProvider {
    static var defaultValue: String { "other" }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/core/model/Billing.kt
package com.loonext.android.core.model

import kotlinx.serialization.Serializable

/** POST /v1/billing/checkout and /portal — open in an EXTERNAL browser. */
@Serializable
data class HostedUrl(val url: String)

/**
 * #583 — a prepaid year, as far as this client needs it.
 *
 * The phones do not SELL a year (that lives on web, where the up-front payment
 * belongs), so this is deliberately not the whole `GET /v1/billing/prepay` shape.
 * What a phone needs is the half that changes what a plan switch does: a year is
 * running, and this much of it comes back if you end it.
 *
 * EVERY FIGURE IS IN `currency`, which is what was COLLECTED — a year bought before
 * the CAD option was filed is genuinely USD even on a workspace billed in CAD today.
 * Printing it in the workspace's current currency would relabel somebody's payment.
 *
 * Defaults everywhere, because this decodes a response written by a newer or older
 * server than the app: an absent `conversion` reads as "no figure to show" and the
 * panel says the year ends without quoting an amount, rather than crashing.
 */
@Serializable
data class PrepaidConversion(
    val consumed_months: Int = 0,
    val credit_cents: Int = 0,
)

@Serializable
data class OpenPrepaidYear(
    val plan: String = "starter",
    val amount_cents: Int = 0,
    val currency: String = "usd",
    val granted_through: String? = null,
    val conversion: PrepaidConversion? = null,
)

@Serializable
data class PrepayOffer(
    val open: OpenPrepaidYear? = null,
)


@Serializable
data class UsageMonth(val month: String, val segments: Long)

@Serializable
data class UsageStorage(
    val attachments_bytes: Long = 0,
    val mms_bytes: Long = 0,
    /** Media a customer sent us. */
    val received_media_bytes: Long = 0,
    /** Media we sent out. */
    val sent_media_bytes: Long = 0,
    /** Voicemail recordings we keep in our own bucket. */
    val voicemail_bytes: Long = 0,
    /** Anything stored that the named kinds do not account for. */
    val other_bytes: Long = 0,
    /** Every byte this workspace holds, measured from the buckets themselves. */
    val total_bytes: Long = 0,
) {
    /**
     * What is really stored. The old line added two figures together and so
     * left voicemail recordings out entirely; the server measures the buckets
     * now, and the sum is only a fallback for a response that predates it.
     */
    val totalStored: Long
        get() = if (total_bytes > 0) total_bytes else attachments_bytes + mms_bytes
}

@Serializable
data class UsageVoice(
    val used_minutes: Long = 0,
    val included_minutes: Long = 0,
    val cap_minutes: Long? = null,
    val overage_minutes: Long = 0,
    val projected_overage_cents: Long = 0,
    val overage_billed: Boolean = true,
)

@Serializable
data class UsageOverageProjection(
    val trending_over: Boolean = false,
    val projected_overage_cents: Long = 0,
)

/**
 * #178: the fair-use presentation contract. GET /v1/usage derives `status`
 * server-side so every client renders the same philosophy: 'quiet' shows no
 * meters anywhere, 'pacing' shows the early warning, 'capped' shows the
 * owner-set spending cap approaching (>=90%) or reached.
 */
object UsageStatus {
    const val QUIET = "quiet"
    const val PACING = "pacing"
    const val CAPPED = "capped"
}

/** GET /v1/usage — nulls when the company has never checked out. */
@Serializable
data class Usage(
    /** #178 presentation status; the default keeps pre-#178 cached payloads
     *  decoding as the calm state (unknown values also render quiet). */
    val status: String = UsageStatus.QUIET,
    /**
     * #522: the currency EVERY *_cents figure on this payload is quoted in.
     *
     * The server states it rather than leaving this client to pair the numbers
     * with `companies.billing_currency` read separately — those two can
     * disagree, and the failure is silent: the screen labels a CAD figure USD.
     * Defaults to USD so a payload from a server that predates it reads as what
     * every workspace was actually charged then.
     */
    val currency: String? = null,
    val period_start: String? = null,
    val period_end: String? = null,
    val included_segments: Long = 0,
    val used_segments: Long = 0,
    val inbound_segments: Long = 0,
    val overage_segments: Long = 0,
    val cap_segments: Long? = null,
    val projected_overage_cents: Long = 0,
    val overage_projection: UsageOverageProjection = UsageOverageProjection(),
    val history: List<UsageMonth> = emptyList(),
    val storage: UsageStorage = UsageStorage(),
    val voice: UsageVoice = UsageVoice(),
    /**
     * What Lou has done this month, per feature. Defaults to empty so a client
     * talking to a server that predates it renders no section rather than
     * failing to decode.
     */
    val ai: List<AiFeatureUsage> = emptyList(),
    /**
     * #426: carrier-reported delivery for the period. Null on a server that
     * predates it, or when the read failed — the rest of the page still shows.
     */
    val delivery: UsageDelivery? = null,
)

/**
 * #426 — carrier-reported delivery, split by where the message was going.
 *
 * The NAME is load-bearing: a receipt means a carrier acknowledged handoff,
 * not that a person read it, so every surface says "carrier-reported".
 */
@Serializable
data class UsageDeliveryCountry(
    /** "US" | "CA" | "other", from the destination's area code. */
    val country: String = "other",
    val delivered: Long = 0,
    val failed: Long = 0,
    /** Accepted by us, not yet acknowledged by a carrier. Not a failure. */
    val pending: Long = 0,
    /**
     * delivered / (delivered + failed), or NULL when too few have settled to
     * mean anything. Render counts and never a percentage when null: one
     * failure out of forty reads as 2.5% and usually means a disconnected
     * number, which is manufactured worry rather than information.
     */
    val rate: Double? = null,
)

@Serializable
data class UsageDelivery(
    val by_country: List<UsageDeliveryCountry> = emptyList(),
    val delivered: Long = 0,
    val failed: Long = 0,
    val pending: Long = 0,
)

/** One AI feature's month: what has been used against its limit. */
@Serializable
data class AiFeatureUsage(
    /** The ledger key, so a row is identified without matching on copy. */
    val key: String,
    val label: String,
    val used: Long = 0,
    val cap: Long = 0,
    val enabled: Boolean = true,
    /**
     * #431 ask 3 — what people did with the output, beside what it cost.
     *
     * Labelled by the server in each feature's own words ("sent as written",
     * "cleared") so all three clients say the same thing. EMPTY until outcomes
     * arrive, and an empty list must render as "not measured yet" rather than as
     * zeroes: a feature used forty times with nothing recorded is an
     * instrumentation gap, and "0 sent as written" would report that gap as a
     * verdict on the quality.
     */
    val outcomes: List<AiOutcomeLine> = emptyList(),
    /**
     * How many outcomes those lines cover. Separate from `used` because they
     * will not match — a draft offered and never read is a request with no
     * outcome — and no rate is computed anywhere, deliberately.
     */
    val outcomesRecorded: Long = 0,
)

/** #431: what a person did with one feature's output, ready to render. */
@Serializable
data class AiOutcomeLine(val label: String, val count: Long = 0)

/** GET /v1/billing/modules — admin-only add-on catalog with enabled state. */
@Serializable
data class BillingModules(val modules: List<BillingModule> = emptyList())

@Serializable
data class BillingModule(
    val id: String,
    val label: String,
    val blurb: String,
    val detail: String? = null,
    val monthly_cents: Long,
    val enabled: Boolean = false,
    /** #41: deliverable AND priced in this environment; refuse to sell otherwise. */
    val available: Boolean = false,
)

/**
 * #523 — a number this workspace holds that its plan does not currently cover.
 *
 * NOT RELEASED, and the word is chosen: the row is still ours, still receiving,
 * and its history is untouched. It simply cannot send or answer. Every surface
 * that renders one of these has to carry that distinction, because "suspended"
 * on its own reads as "gone" to the person whose van has the number on it.
 */
@Serializable
data class HeldNumber(
    val id: String,
    val number_e164: String? = null,
    /** When it went on hold. Null for a row suspended before #523 shipped. */
    val suspended_at: String? = null,
)

/**
 * #523 — why a number is on hold, as the SERVER decides it.
 *
 * Two values, not three, and the split is the one thing this client must not
 * re-derive: `over_plan_allowance` means the subscription is live and the
 * workspace simply holds more numbers than it pays for, which is the state with
 * a way out; `subscription_inactive` means the suspension belongs to the
 * cancellation or the failed payment, which other cards on the same screen
 * already own. Rendering both from one card would be two explanations of one
 * suspension sitting an inch apart.
 */
object HeldNumberReason {
    const val OVER_PLAN_ALLOWANCE = "over_plan_allowance"
    const val SUBSCRIPTION_INACTIVE = "subscription_inactive"
}

/**
 * GET /v1/billing/held-numbers (#523) — owner/admin only (`billing.manage`).
 *
 * EVERY FIGURE ON THIS CARD ARRIVES HERE. The allowance, the plan's hard cap
 * and the price of un-holding one are served rather than derived, and
 * [extra_number_cents] carries [extra_number_currency] beside it because the
 * extra-number price book is filed in USD only — a bare "$5" in front of a
 * Canadian reader means CAD and is the #522 defect verbatim.
 *
 * Defaults everywhere, so a client talking to a Worker that predates the route
 * decodes an empty state and draws nothing rather than failing the read.
 */
@Serializable
data class HeldNumbers(
    /** "starter" | "pro" | null (no checkout yet). */
    val plan: String? = null,
    /** Numbers the plan itself covers. Null when [plan] is null. */
    val included: Int? = null,
    /** Extra-number capacity actually billed today. */
    val paid_extras: Int = 0,
    /** included + paid_extras — what may be live at once. Null with no plan. */
    val allowance: Int? = null,
    /** The plan's hard TOTAL cap (#80), or null when it has none (Pro). */
    val max_total: Int? = null,
    /** [HeldNumberReason], or null when nothing is held. */
    val reason: String? = null,
    val held: List<HeldNumber> = emptyList(),
    /** What buying capacity for ONE held number costs. */
    val extra_number_cents: Int? = null,
    /** The currency [extra_number_cents] is denominated in ("usd"). */
    val extra_number_currency: String? = null,
    /**
     * Whether the reinstate route would be accepted right now. Served so the
     * button can be ABSENT rather than fail — being told "no" after pressing it
     * is how somebody concludes the product is broken.
     */
    val can_reinstate: Boolean = false,
    /** Starter only: the other way back, which buys no extra number. */
    val can_upgrade: Boolean = false,
)

/** POST /v1/billing/held-numbers/:id/reinstate (#523). */
@Serializable
data class ReinstateResult(
    val reinstated: Boolean = false,
    /**
     * It was already back — a double-press, or an upgrade reinstated it between
     * the card painting and the button being pressed. Not an error, and
     * emphatically not a second charge.
     */
    val already_active: Boolean = false,
    val paid_extras: Int? = null,
    val allowance: Int? = null,
    val held: List<HeldNumber> = emptyList(),
)

/** POST /v1/billing/change-plan result. */
@Serializable
data class ChangePlanResult(
    val plan: String,
    val effective: String,
    val effective_at: String? = null,
    /**
     * #523: what the bigger allowance brought back with it. Pro includes two
     * numbers, so an upgrade is one of the two routes out of a hold — and an
     * owner who pays $79 to get their second line back needs to be told it
     * worked rather than left to go and look. Empty on every ordinary upgrade.
     */
    val reinstated: List<HeldNumber> = emptyList(),
    /** Still on hold after the upgrade — a Pro workspace can hold three. */
    val held: List<HeldNumber> = emptyList(),
)
